bundlegen/core/bundle_processor.py

```diff
@@ -536,29 +537,175 @@ def _process_logging(self):
+            if (cnt > max):
+                libc = lib
+                max = cnt
+
+        if (libc is None):
+            return
+
+        logger.debug(f"Found libc: {libc['name']}")
+        sublibs.remove(libc)
+        libc['sublibs'] = []
+        for sublib in sublibs:
+            libc['sublibs'].append(sublib['name'])
+            sublib['parentlib'] = libc['name']
+            sublib['apiversions'] = []
+
+    def _take_rootfs_lib(self, dstlib, api_info):
+        """ The lib version from image rootfs was choosen. Basically log here that it was choosen.
+            Also for any sublibs if present.
+        """
+        logger.debug(f"IMAGE version choosen: {dstlib}")
+        self.handled_libs.add(dstlib)
+        if api_info and api_info.get('sublibs'):
+            for sublib in api_info['sublibs']:
+                logger.debug(f"IMAGE version choosen: {sublib}")
+                self.handled_libs.add(sublib)
+
+    # ==========================================================================
+    def _mount_or_use_rootfs(self, srclib, dstlib):
+        """Determine to mount lib from host OR use the one inside bundle rootfs.
+           If lib exists in rootfs and apiversions info exists for it inside *_libs.json config
+           then we try to use that version of the lib that has most API versions defined inside.
+           If lib does not exist inside image rootfs then create mount bind.
+           If no API versions info exists then create mount bind.
+           If decision cannot be made because API versions disjunct, choose image rootfs version.
+
+        Args:
+            srclib (string): libpath on host. For this lib, the api info will be looked up
+                             inside *_libs.json.
+            dstlib (string): libpath in rootfs image. In most cases the same as srclib. For this lib
+                             the api info will be read directly from the lib file via readelf.
+        """
+        if dstlib in self.handled_libs:
+            #logger.debug(f"Already handled: {dstlib}")
+            return
+
+        api_info = None
+        if self.platform_cfg.get('libs'):
+            api_info = [x for x in self.platform_cfg['libs'] if x['name'] == srclib]
+        if not api_info:
+            logger.debug(f"No api info found for {dstlib}")
+            self._take_host_lib(srclib, dstlib, None)
+            return
+
+        api_info = api_info[0]
+
+        # if this is a sublib then switch to logic of parentlib instead
+        if api_info.get('parentlib'):
+            self._mount_or_use_rootfs(api_info['parentlib'], api_info['parentlib'])
+            return
+
+        rootfs_filepath = os.path.join(self.rootfs_path, dstlib.lstrip('/'))
+        if not os.path.exists(rootfs_filepath):
+            logger.debug(f"Lib not inside image rootfs {dstlib}")
+            self._take_host_lib(srclib, dstlib, api_info)
+            return
+
+        if len(api_info['apiversions']) > 0:
```

Comment:
This just compares which library has more API versions. Is it possible for the host version to have apiVersions `[v0.1, v0.2]` and the image to have `[v1.0]`? In that scenario, the host library with two versions would win here, even though the second is a newer API version (and they're likely incompatible)

Not sure if you know if this is likely to happen but would like to hear your thoughts.

Reply:
Actually, in such case the current algorithm cannot decide because [v0.1, v0.2] is not a subset of [v1.0] and also not the other way around. Current algorithm doesn't try to parse the actual version strings. I suppose this could be added in a next iteration/change. In case the algo cannot decide it currently chooses to prefer the dac rootfs version of the lib.

Current implementation seems to handle the libc case correctly since newer versions of libc seem to add an api string for each new version bump (and keeping the older api version strings). So the "subset" checking mechanism works.